internals/daemon/daemon.go

```diff
@@ -747,9 +747,34 @@ func (d *Daemon) Stop(sigCh chan<- os.Signal) error {
 	return nil
 }
 
-// stopRunningServices stops all running services, waiting for a short time
-// for them all to stop.
+// stopRunningServices waits for all running changes to complete then stops all running services,
+// waiting for a short time for them all to stop.
 func (d *Daemon) stopRunningServices() error {
+	st := d.state
+	st.Lock()
+	var runningChanges []*state.Change
+	for _, chg := range st.Changes() {
+		status := chg.Status()
+		if status == state.DoStatus || status == state.DoingStatus ||
+			status == state.UndoStatus || status == state.UndoingStatus {
+			runningChanges = append(runningChanges, chg)
+		}
+	}
+	st.Unlock()
+
+	if len(runningChanges) > 0 {
+		logger.Noticef("Waiting for %d running changes to complete before stopping services.", len(runningChanges))
+		for _, chg := range runningChanges {
+			select {
+			case <-chg.Ready():
+				logger.Debugf("Change %s completed.", chg.ID())
+			case <-time.After(d.overlord.ServiceManager().StopTimeout()):
```

Comment:
We should only call `time.After` once before the for loop, so it's not StopTimeout times the number of changes, but only one timeout overall.

```go
timeout := time.After(d.overlord.ServiceManager().StopTimeout())
for _, chg := range runningChanges {
    select {
    case <-chg.Ready(): ...
    case <-timeout: ...
    }
}
```